pulp_cli/__init__.py

```diff
@@ -1,38 +1,208 @@
+import os
 import sys
 import typing as t
+import warnings
 from types import ModuleType
 
 import click
+import toml
+from pulp_glue.common.i18n import get_translation
+
+from pulp_cli.config import CONFIG_LOCATIONS, config, config_options, validate_config
+from pulpcore.cli.common.generic import PulpCLIContext, pulp_group
 
 if sys.version_info >= (3, 10):
     from importlib.metadata import entry_points
 else:
     from importlib_metadata import entry_points
 
-__version__ = "0.24.0.dev"
-_main: t.Optional[click.Group] = None
+try:
+    from click_shell import make_click_shell
+
+    HAS_CLICK_SHELL = True
+except ImportError:
+    HAS_CLICK_SHELL = False
 
+__version__ = "0.24.0.dev"
+translation = get_translation(__package__)
+_ = translation.gettext
+# Keep track to prevent loading plugins twice
+loaded_plugins: t.Set[str] = set()
 
-def load_plugins() -> click.Group:
-    global _main
 
+def load_plugins(enabled_plugins: t.Optional[t.List[str]] = None) -> None:
     ##############################################################################
     # Load plugins
     # https://packaging.python.org/guides/creating-and-discovering-plugins/#using-package-metadata
+
     discovered_plugins: t.Dict[str, ModuleType] = {
         entry_point.name: entry_point.load()
         for entry_point in entry_points(group="pulp_cli.plugins")
+        if (enabled_plugins is None or entry_point.name in enabled_plugins)
+        and entry_point.name not in loaded_plugins
     }
-    _main = discovered_plugins["common"].main
-    assert isinstance(_main, click.Group)
-    for plugin in discovered_plugins.values():
+    for name, plugin in discovered_plugins.items():
         if hasattr(plugin, "mount"):
-            plugin.mount(_main, discovered_plugins=discovered_plugins)
-    return _main
+            plugin.mount(main, discovered_plugins=discovered_plugins)
+        else:
+            warnings.warn(
+                DeprecationWarning("Plugin '{name}' has no `mount` entrypoint which is deprecated.")
+            )
+        loaded_plugins.add(name)
+
+
+##############################################################################
+# Main entry point
+
+
+CONFIG_KEY = f"{__name__}.config"
+PROFILE_KEY = f"{__name__}.profile"
+
+
+def _config_callback(ctx: click.Context, param: t.Any, value: t.Optional[str]) -> None:
+    ctx.meta[CONFIG_KEY] = value
+    if PROFILE_KEY in ctx.meta:
+        _load_config(ctx)
```

Comment:
Maybe these deserve a comment to how these callbacks work/called, especially since they now control how the plugins are loaded? I think one thing that confused me is that while both callbacks are eager, the order they run in will depend:

1. The order they are specified on the command line
2. Otherwise the order they are defined in on the command
https://click.palletsprojects.com/en/8.1.x/advanced/#callback-evaluation-order

The logic of these callbacks are not easily understood without some deep knowledge of click callbacks, so a comment would be helpful for future reference.

Reply:
Maybe i can improve that.
The basic idea is They are both eager (running before everything else), they both run in any circumstances, and both of them only act when the other one already deposited it's value in the `ctx.meta` store. Therefor, order does not matter.